### PR TITLE
Bashwizard integration + docker-free test runner

### LINKED ISSUE #56: As a developer, I want to run integration tests for each CI build run
### Description
*As a developer, I want to run integration tests for each CI build run*
Running end to end integration tests to validate terraform deployments are operating as expected is essential for enterprises. Similar to bedrock, we'll use [terratest ](https://github.com/gruntwork-io/terratest)in the spirit of following azure devops [guidance](https://docs.microsoft.com/en-us/azure/terraform/terratest-in-terraform-modules) for terraform deployments. Initially, we'll run a light weight E2E test similar to bedrock's [azure-simple](https://github.com/Microsoft/bedrock/blob/master/test/bedrock_Azure_simple_test.go#L36) integration test. 

Creating the terraform template is out of scope for this feature. We'll use the app service terraform hello world sample [here](https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/examples/app-service/linux-basic/main.tf) as our temporary test bed placeholder until the `azure-simple` template is released. This integration test should run terraform init, plan, format, destroy and apply on the mocked release and assert that it can establish a connection to the provisioned app service website with a 200 response.  

This azure devops build step will get triggered for pull request updates/submissions. The integration test harness should support both local and cloud runs. To simplify local dependency setup we'll use the docker base image from feature #55. We'll need to add a docker build step to the pipeline as the integration test(s) will run as a container.

### Acceptance Criteria
- [ ] Should assert that we have a mocked terraform template which our int test can apply which scaffolds a hello world app service website. 
- [ ] Should assert that we have a go terratest module that acts as an integration test which validates the response status code of the provisioned app service website.  
- [ ] Should assert that the go test run's within a docker container using `mage` and runs successfully locally and in azure devops. 
- [ ] Should assert that the `README` in the `test` directory is updated with local and cloud setup instructions.  

Also, here are a few points that need to be addressed:

1. Issues #55 and #54 are pre-req's

### Resources
*Mockups*
Include an animated GIF of the build running successfully in azDO

### Tasks
*Add the go test harness module to the `test` folder
*Add a mocked app service website terraform template within `environments`. This is a temporary solution to validate the test module until `azure-simple` is released. 
*Update the documentation in the test folder. 
*Dockerize the test harness run

### PR DESCRIPTION
**Pull request by erikschlegel**
## All Submissions:
-------------------------------------
* [YES] Have you followed the guidelines in our Contributing [document](./CONTRIBUTING.md)?
* [YES] Have you added an explanation of what your changes do and why you'd like us to include them?
* [YES] I have updated the documentation accordingly.
* [NA] I have added tests to cover my changes.
* [YES] All new and existing tests passed.
* [YES] My code follows the code style of this project.
* [YES] I ran lint checks locally prior to submission.
* [YES] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

Issue Number: closes #56 

## What is the new behavior?
-------------------------------------

***This pull request covers the following feature additions.***
- Updated test harness getting started and shell script argument options documentation.
- Refactored all our shell scripts to adhere to the scaffolding format from @joelong01 [BashWizard](https://github.com/joelong01/BashWizard) project. 
- Added support for a docker free version of the test harness to minimize local build time for active development scenarios. 
- Added support for `template_name_override` option for both test harness runners. This allows developers to override the target template folder for automated test runs.
- Added manifest files to `.gitignore`
- Removed overlapping / unused module folders as we're referencing `vnet`, `trafficmanager` from bedrock going forward. 

## Does this introduce a breaking change?
-------------------------------------
- [YES]

## Any relevant logs, error output, etc?
-------------------------------------
***Local build log for passed test run following change***
```
TestITAzureSimple 2019-05-06T17:01:01-05:00 command.go:121: ?[0m?[1mazurerm_resource_group.main: Still destroying... (ID: /subscriptions/49e3a7a8-c63e-4124-948c-ee51b74d5801/resourceGroups/cobalt-MM8WtJ, 20s elapsed)?[0m?[0m
TestITAzureSimple 2019-05-06T17:01:11-05:00 command.go:121: ?[0m?[1mazurerm_resource_group.main: Still destroying... (ID: /subscriptions/49e3a7a8-c63e-4124-948c-ee51b74d5801/resourceGroups/cobalt-MM8WtJ, 30s elapsed)?[0m?[0m
TestITAzureSimple 2019-05-06T17:01:21-05:00 command.go:121: ?[0m?[1mazurerm_resource_group.main: Still destroying... (ID: /subscriptions/49e3a7a8-c63e-4124-948c-ee51b74d5801/resourceGroups/cobalt-MM8WtJ, 40s elapsed)?[0m?[0m
TestITAzureSimple 2019-05-06T17:01:27-05:00 command.go:121: ?[0m?[1mazurerm_resource_group.main: Destruction complete after 46s?[0m?[0m
TestITAzureSimple 2019-05-06T17:01:28-05:00 command.go:121: ?[0m?[1m?[32m
TestITAzureSimple 2019-05-06T17:01:28-05:00 command.go:121: Destroy complete! Resources: 3 destroyed.?[0m
--- PASS: TestITAzureSimple (209.33s)
PASS
ok      cobalt/build/azure-simple-hw/test/integration   209.467s
/c/Users/erisch/go/src/cobalt
```

### CODE REVIEW

**Comment by joelong01** [comment]
Yay! 😊

**Comment by NathanielRose** [comment]
Looks like the [msftcse image](https://github.com/microsoft/cobalt/pull/81/files#diff-389656a19bda4636b1c0c52a8281822aR46) is private on docker hub.

**Comment by NathanielRose** [comment]
Running the docker image out of box and receiving an error:
```
Successfully tagged msftcse/cobalt-test-base:g1.11.5t0.11.13
nmrose@MININT-IQPVRH0:/mnt/c/Users/naros/Desktop/FY19/cobalt/cobalt$
nmrose@MININT-IQPVRH0:/mnt/c/Users/naros/Desktop/FY19/cobalt/cobalt$ ./test-harness/local-run.sh
INFO: loading environment
INFO: Checking required environment variables
INFO: passed environment variable check
local-run.sh:
t0.11.13er_base_image_name...................... msftcse/cobalt-test-base:g1.11.5
    tf_state_storage_acct...... cobaltfstate
    tf_state_container.... cobaltfstate-remote-state-container
    template_name_override....
INFO: verified that environment is fully defined
INFO: verified that environment is fully defined
INFO: Running git diff from branch (HEAD
fatal: ambiguous argument '(HEAD': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
INFO: No templates to process. Exiting build step
```

**Comment by NathanielRose** [comment]
Local running of the script gives the following output:
```
nmrose@MININT-IQPVRH0:/mnt/c/Users/naros/Desktop/FY19/cobalt/cobalt$ ./test-harness/local-run-wo-docker.sh
INFO: loading environment
INFO: Checking required environment variables
INFO: passed environment variable check
local-run-wo-docker.sh:
    tf_state_storage_acct...... cobaltfstate
    tf_state_container.... cobaltfstate-remote-state-container
    template_name_override....
INFO: verified that environment is fully defined
INFO: Setting up golang manifest and vendor packages
init failed: unable to detect the containing GOPATH: /mnt/c/Users/naros/Desktop/FY19/cobalt/cobalt is not within a known GOPATH/src
```

**Comment by erikschlegel** [comment]
> Looks like the [msftcse image](https://github.com/microsoft/cobalt/pull/81/files#diff-389656a19bda4636b1c0c52a8281822aR46) is private on docker hub.

The public image can be found in the `msftcse/cobalt-test-base` repo [here](https://hub.docker.com/r/msftcse/cobalt-test-base). Updated the documentation to be more clear.

**Comment by erikschlegel** [comment]
> Running the docker image out of box and receiving an error:
> 
> ```
> Successfully tagged msftcse/cobalt-test-base:g1.11.5t0.11.13
> nmrose@MININT-IQPVRH0:/mnt/c/Users/naros/Desktop/FY19/cobalt/cobalt$
> nmrose@MININT-IQPVRH0:/mnt/c/Users/naros/Desktop/FY19/cobalt/cobalt$ ./test-harness/local-run.sh
> INFO: loading environment
> INFO: Checking required environment variables
> INFO: passed environment variable check
> local-run.sh:
> t0.11.13er_base_image_name...................... msftcse/cobalt-test-base:g1.11.5
>     tf_state_storage_acct...... cobaltfstate
>     tf_state_container.... cobaltfstate-remote-state-container
>     template_name_override....
> INFO: verified that environment is fully defined
> INFO: verified that environment is fully defined
> INFO: Running git diff from branch (HEAD
> fatal: ambiguous argument '(HEAD': unknown revision or path not in the working tree.
> Use '--' to separate paths from revisions, like this:
> 'git <command> [<revision>...] -- [<file>...]'
> INFO: No templates to process. Exiting build step
> ```

can you provide the output of `git branch`?

**Comment by erikschlegel** [comment]
> Local running of the script gives the following output:
> 
> ```
> nmrose@MININT-IQPVRH0:/mnt/c/Users/naros/Desktop/FY19/cobalt/cobalt$ ./test-harness/local-run-wo-docker.sh
> INFO: loading environment
> INFO: Checking required environment variables
> INFO: passed environment variable check
> local-run-wo-docker.sh:
>     tf_state_storage_acct...... cobaltfstate
>     tf_state_container.... cobaltfstate-remote-state-container
>     template_name_override....
> INFO: verified that environment is fully defined
> INFO: Setting up golang manifest and vendor packages
> init failed: unable to detect the containing GOPATH: /mnt/c/Users/naros/Desktop/FY19/cobalt/cobalt is not within a known GOPATH/src
> ```

Please confirm that all prereq [conditions](https://github.com/microsoft/cobalt/tree/erisch/features/bashwizard/test-harness#prerequisites-1) are accounted for?

**Comment by NathanielRose** [comment]
> > Running the docker image out of box and receiving an error:
> > ```
> > Successfully tagged msftcse/cobalt-test-base:g1.11.5t0.11.13
> > nmrose@MININT-IQPVRH0:/mnt/c/Users/naros/Desktop/FY19/cobalt/cobalt$
> > nmrose@MININT-IQPVRH0:/mnt/c/Users/naros/Desktop/FY19/cobalt/cobalt$ ./test-harness/local-run.sh
> > INFO: loading environment
> > INFO: Checking required environment variables
> > INFO: passed environment variable check
> > local-run.sh:
> > t0.11.13er_base_image_name...................... msftcse/cobalt-test-base:g1.11.5
> >     tf_state_storage_acct...... cobaltfstate
> >     tf_state_container.... cobaltfstate-remote-state-container
> >     template_name_override....
> > INFO: verified that environment is fully defined
> > INFO: verified that environment is fully defined
> > INFO: Running git diff from branch (HEAD
> > fatal: ambiguous argument '(HEAD': unknown revision or path not in the working tree.
> > Use '--' to separate paths from revisions, like this:
> > 'git <command> [<revision>...] -- [<file>...]'
> > INFO: No templates to process. Exiting build step
> > ```
> 
> can you provide the output of `git branch`?

nmrose@MININT-IQPVRH0:/mnt/c/Users/naros/Desktop/FY19/cobalt/cobalt$ git branch
` (HEAD detached at origin/erisch/features/bashwizard)` :slightly_smiling_face:

File: test-harness/README.md [hunk 1]
```diff
@@ -10,46 +10,127 @@ In addition, the cobalt test suite allows for better collaboration with embeddin
...
+
+    ```script
+    $ echo $GOPATH
+    C:\Users\erisch\go
+    $ pwd
+    /c/Users/erisch/go/src/cobalt
+    ```
+
+- Install [Azure CLI](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest)
+- Install golang's dep package manager via Git Bash.
```
Comment (nmiodice) [inline test-harness/README.md]: Probably should not assume git bash here, as even on windows people may choose to install this with WSL.

Comment (erikschlegel) [reply test-harness/README.md]: the test harness has also been tested on wsl

File: test-harness/README.md [hunk 2]
```diff
@@ -10,46 +10,127 @@ In addition, the cobalt test suite allows for better collaboration with embeddin
...
+        go version  : go1.10.3
+        go compiler : gc
+        platform    : windows/amd64
+        features    : ImportDuringSolve=false
+    ```
+
+- Install [Terraform](https://learn.hashicorp.com/terraform/getting-started/install.html)
+
+#### Local Run Script
+
```
[the remark this answers is not in the document]
Comment (erikschlegel) [reply test-harness/README.md]: `*.toml` has been added to the `.gitignore`

File: .env.template [hunk 3]
```diff
@@ -6,4 +6,6 @@ ARM_CLIENT_ID="<az-service-principal-client-id>"
 ARM_CLIENT_SECRET="<az-service-principal-auth-secret>"
 ARM_TENANT_ID="<az-service-principal-tenant>"
 ARM_ACCESS_KEY="<remote-state-storage-account-primary-key>"
```
[the remark this answers is not in the document]
Comment (erikschlegel) [reply .env.template]: Terraform references the `ARM_ACCESS_KEY` env var when capturing remote deployment state.

**Comment by erikschlegel** [comment]
> LGTM! - Will eventually want to further investigate go.mod to move outside go/src.

Thanks for the review @NathanielRose . Let us know how we can help with migrating from `dep` -> `go.mod` 🚀

**Pull request merged**